src/collapse/xdebug.rs

```diff
@@ -0,0 +1,367 @@
+    where
+        R: BufRead,
+        W: Write,
+    {
+        // Timings for all call stacks in total.
+        let mut stacks: HashMap<_, f32> = HashMap::new();
+        let mut current_stack = CallStack::new();
+        let mut prev_start_time = 0.0;
+        let mut line = String::new();
+
+        loop {
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with(TRACE_START) {
+                break;
+            }
+
+            line.clear();
+        }
+
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with(TRACE_END) {
+                break;
+            }
+
+            // Break the line into tab separated tokens.
+            let mut parts = Fields::new(&line).skip(2);
+
+            let (is_exit, time) = if let (Some(is_exit), Some(time)) = (parts.next(), parts.next()) {
+                let is_exit = match is_exit {
+                    "1" => true,
+                    "0" => false,
+                    a => panic!(format!("uh oh: {}", a)),
+                };
+
+                let time = time.parse::<f32>().unwrap();
+
+                (is_exit, time)
+            } else {
+                continue;
+            };
+
+            if is_exit && current_stack.is_empty() {
+                warn!("Found function exit without corresponding entrance. Discarding line. Check your input.");
+                continue;
+            }
+
+            {
+                let current = current_stack.current();
+                let duration = SCALE_FACTOR * (time - prev_start_time);
+
+                // hash `current` for lookup and insertion.
+                let mut hasher = stacks.hasher().build_hasher();
+                current.hash(&mut hasher);
+                let hash = hasher.finish();
+
+                match stacks
+                    .raw_entry_mut()
+                    .from_key_hashed_nocheck(hash, current)
+                {
+                    RawEntryMut::Occupied(mut occ) => *occ.get_mut() += duration,
+                    RawEntryMut::Vacant(vacant) => {
+                        // `Box<str>` has same hash as `str`.
+                        vacant.insert_hashed_nocheck(hash, Box::from(current), duration);
+                    }
+                }
+            }
+
+            if is_exit {
+                current_stack.pop();
+            } else {
+                let _ = parts.next();
+                let func_name = parts.next();
+                let _ = parts.next();
+                let path_name = parts.next();
+
+                if let (Some(func_name), Some(path_name)) = (func_name, path_name) {
+                    current_stack.call(func_name, path_name);
+                }
+            }
+
+            prev_start_time = time;
+        }
+
+        for (key, value) in stacks {
```

Comment:
Even though it's sad, I think you should sort these like we do in the other collapsers:
https://github.com/jonhoo/inferno/blob/65a2f20246611229f35b8f701c386dbb9531b8e7/src/collapse/perf.rs#L397-L405

We sort because that's what upstream does, and because flamegraph requires its input to be sorted (and re-sorting is faster if it's already sorted).